src/ceph-volume/ceph_volume/util/disk.py

```diff
@@ -170,6 +170,23 @@ def device_family(device):
     return devices
 
 
+def udevadm_property(device, properties=[]):
+    """
+    Query udevadm for information about device properties.
+    Optionally pass a list of properties to return. A requested property might
+    not be returned if not present.
+    """
+    cmd = ['udevadm', 'info', '--query=property', device]
+    out, _err, _rc = process.call(cmd)
+    ret = {}
+    for line in out:
+        p, v = line.split('=', 1)
```

Comment:
this should catch a `TypeError` as that can be the case when the split doesn't have the two items being assigned here.

I would recommend splitting the system call with a parser, similar to what we've done with `blkid` and `lsblk` because it is easier to test the parsing vs. mixing it with the system call (this PR needs a few of those to ensure that this function is working as expected). 

Curious that you have implemented this, I'm working on removing partitions as part of `lvm zap` which would need `udevadm`, here is where I left that implementation at:

    def udevadm_info(device):
        """
        Capture device information with udevadm, useful in situations where blkid and lsblk aren't
        producing enough useful output.
        """
        command = ['udevadm', 'info', '--query=property', '--name', device]
        out, err, rc = process.call(command)
        return _udevadm_parser(out)
    
    
    def _udevadm_parser(output):
        """
        Parses the output from a udevadm call, requires output to be
        produced using the ``info``, specifying the actual --query type, to remove prefixes in the output.
    
        Expected output::
    
            # udevadm info --query=property --name /dev/sdb
            DEVLINKS=/dev/disk/by-path/pci-0000:00:14.0-scsi-0:0:3:0
            DEVNAME=/dev/sdb
            DEVPATH=/devices/pci0000:00/0000:00:14.0/host3/target3:0:3/3:0:3:0/block/sdb
            DEVTYPE=disk
            ID_BUS=scsi
            ID_MODEL=HARDDISK
            ID_MODEL_ENC=HARDDISK\x20\x20\x20\x20\x20\x20\x20\x20
            ID_PART_TABLE_TYPE=gpt
            ID_PART_TABLE_UUID=bea7a874-c406-46b2-98bd-414f062f2eb5
            ID_PATH=pci-0000:00:14.0-scsi-0:0:3:0
            ID_PATH_TAG=pci-0000_00_14_0-scsi-0_0_3_0
            ID_REVISION=1.0
            ID_SCSI=1
            ID_TYPE=disk
            ID_VENDOR=VBOX
            ID_VENDOR_ENC=VBOX\x20\x20\x20\x20
            MAJOR=8
            ...
        """
        processed = {}
        for line in output:
            try:
                key, value = line.split('=')
            except ValueError:
                continue
            processed[key] = value
        return processed

Reply:
I would strongly argue against catching an error here.
1) It's highly unlikely the output of udevadm is ever going to change.
2) You propose to silently eat an error in the udevadm output. What is the benefit of that? Fail early is imho much better here as this is an error (udevadm giving output that is not dividable at a ```=```) that seems impossible to recover from. Catching exceptions is meant for error handling, not error hiding.

Reply:
The output of udevadm might not change at all for a while, but the parser must be able to withstand lines that don't comply exactly to what it needs. 

There is no actual error eating that is silent here, as all system calls that fail are logged to the configured file (would be caused by a non-zero exit status). This is one of the benefits of implementing a parser separately: it doesn't care about error handling from a non-zero exit status, makes testing far easier (just a list of strings is needed).

If we wanted to fail hard when udevadm is not giving us exactly what we need because of incomplete/invalid data or a non-zero exit status we could improve the handling in the function that does the system call.

My example is crufty (work in progress) and you are catching nits that need to be improved for sure! The main idea is to separate the concerns for easier/better unit testing.

Reply:
My 2 cents:
- My reading of the code posted by @alfredodeza continuing after `ValueError` is that I would expect that code to silently eat the failure if the output changes, which also seems a bad idea to me. If that is not the actual behavior of the code, then it is obscured from me, and I believe that section can be improved to be more clear about the real-world behavior.
- Separation of concerns (i.e., parsing output separate from calling udevadm) does seem like a good thing to do for unit testing, and I like that the expected output is spelled out clearly in Alfredo's example. It may be verbose, but as a newcomer to this codebase, it's incredibly helpful to me.
- Allowing the parser to withstand lines that don't exactly match the expected input seems a double-edged sword to me. It may be able to continue functioning in the case of some output change, and it also may begin to function erroneously in an unexpected way in such a case. In my opinion, matching the line exactly is a better solution, as it will be much easier to debug an error at the location of the parser failure than to debug an error which is caused later in the program's lifetime due to an invalid value being parsed and passed on.

Reply:
@BlaineEXE I think we are in agreement :) Like I mentioned, the patch should not be taken as final/production quality because it is a work-in-progress on another branch that hasn't gone out for review.

I agree that parsing shouldn't eat things up (although failed udev calls are logged), and logging them sounds reasonable. 

In short, my ask is to separate concerns for better unit testing, just like the other system calls/functions that exist in that module.

Reply:
> The output of udevadm might not change at all for a while, but the parser must be able to withstand lines that don't comply exactly to what it needs.
> 
> There is no actual error eating that is silent here, as all system calls that fail are logged to the configured file (would be caused by a non-zero exit status). This is one of the benefits of implementing a parser separately: it doesn't care about error handling from a non-zero exit status, makes testing far easier (just a list of strings is needed).

@alfredodeza We are starting to mix issues here. A non-zero exit status is (and should not) be handled here. That is what ```process.call``` does. The issue here is udevadm returning something other then $KEY=$VALUE with a zero exit status. My code will raise an IndexError (in case split returns a list of length 1...i.e. the output has no ```=```)  and I think its bad to catch this and _not handle_ the error. If we could handle the error so the code can continue I'd be all for it, but I don't see a way to handle an unknown udevadm return format.

> If we wanted to fail hard when udevadm is not giving us exactly what we need because of incomplete/invalid data or a non-zero exit status we could improve the handling in the function that does the system call.

How and why would ```process.call``` care about the output format of udevadm? We could add a function parameter to call so it does call an injected parse method, but that would not change the fact, that we can not recover from udevadm returning an unknown format.

@BlaineEXE sees my point I think:

> Allowing the parser to withstand lines that don't exactly match the expected input seems a double-edged sword to me. It may be able to continue functioning in the case of some output change, and it also may begin to function erroneously in an unexpected way in such a case. In my opinion, matching the line exactly is a better solution, as it will be much easier to debug an error at the location of the parser failure than to debug an error which is caused later in the program's lifetime due to an invalid value being parsed and passed on.

Reply:
Oh certainly, a function split makes sense. I was really just arguing about error handling. And no worries, I like a good techinical discussion.